src/Npgsql/Internal/NpgsqlWriteBuffer.cs

```diff
@@ -470,6 +470,31 @@ static async Task WriteBytesLong(NpgsqlWriteBuffer buffer, bool async, byte[] by
         }
     }
 
+    public async Task WriteStreamRaw(Stream stream, int count, bool async, CancellationToken cancellationToken = default)
+    {
+        do
+        {
+            if (WriteSpaceLeft == 0)
+                await Flush(async, cancellationToken);
+            try
+            {
+                var read = async
+                    ? await stream.ReadAsync(Buffer, WritePosition, Math.Min(WriteSpaceLeft, count), cancellationToken)
+                    : stream.Read(Buffer, WritePosition, Math.Min(WriteSpaceLeft, count));
+                if (read == 0)
+                    throw new EndOfStreamException();
+                WritePosition += read;
+                count -= read;
+            }
+            catch (Exception e)
+            {
+                throw Connector.Break(new NpgsqlException("Exception while writing to stream", e));
+            }
+        }
+        while (count > 0);
```

Comment:
If the count argument is 0 (we should support this), I think this code will fail. I think that it you turn the do/while into a regular while, the up-front check would handle that.